language/move-model/src/model.rs

```diff
@@ -1507,11 +1536,25 @@ impl GlobalEnv {
             .map_or_else(|| self.unknown_loc(), |info| info.loc.clone())
     }
 
+    /// Gets the number operation of the given node.
+    pub fn get_node_num_oper(&self, node_id: NodeId) -> NumOperation {
+        self.get_node_num_oper_opt(node_id)
+            .expect("node number oper defined")
+    }
+
     /// Gets the type of the given node.
     pub fn get_node_type(&self, node_id: NodeId) -> Type {
         self.get_node_type_opt(node_id).expect("node type defined")
     }
```

Comment:
This code snippet makes me wonder why we are not fusing the `NumOperation` into `Type`.

Here is the reasoning:
- In Move Model, every expression has a `NodeId` which is associated with `Type`.
- Now in this PR you are introducing a `NumOperation` and tries to associate some `NodeId` with it
- A natural reaction will be, can we merge `NumOperation` with `Type`? Or, the other way round, whether a number is a bitvec or a math number is a *typing* information, it makes perfect sense to me to merge them. So the question becomes: what is the concern that prevent you from merging them?

In fact, I am looking at the `Type` enum in Move Model and one way I see is the following:

```diff
pub enum PrimitiveType {
    Bool,
    U8,
    U16,
    U32,
    U64,
    U128,
    U256,
    Address,
    Signer,
    // Types only appearing in specifications
    Num,
+   BV8,
+   BV16,
+   BV32,
+   BV64,
+   BV128,
+   BV256,
+   NumOrBV // not sure whether we need this bottom case though
    Range,
    EventStore,
}
```

This seems like a more elegant and compact representation of bitvec, plus you get to re-use most of the utilities we have around `Type`, e.g., type unification etc.

Reply:
Perhaps I need to add some clarification on the meanings of the types:

```rust
pub enum PrimitiveType {
    Bool,
    U8,    // bounded integer allowed in arithmetic operations only
    U16,   // bounded integer allowed in arithmetic operations only
    U32,   // bounded integer allowed in arithmetic operations only
    U64,   // bounded integer allowed in arithmetic operations only
    U128,  // bounded integer allowed in arithmetic operations only
    U256,  // bounded integer allowed in arithmetic operations only
    Address,
    Signer,
    // Types only appearing in specifications
    Num,   // unbounded integer allowed in arithmetic operations only
+   BV8,   // fixed-width bitvector allowed in bitwise operations only
+   BV16,  // fixed-width bitvector allowed in bitwise operations only
+   BV32,  // fixed-width bitvector allowed in bitwise operations only
+   BV64,  // fixed-width bitvector allowed in bitwise operations only
+   BV128, // fixed-width bitvector allowed in bitwise operations only
+   BV256, // fixed-width bitvector allowed in bitwise operations only
    Range,
    EventStore,
}
```

NOTES:
- bitshift operators (`<<` and `>>`) are permitted as both arithmetic and bitwise operations. As arithmetic operations, bitshift are simply multiplications and divisions (with modulo operation)
- cast operations are permitted as both arithmetic and bitwise operations.
- an expression must be one of the following cases
  - bounded integer (i.e., `U<N>`)
  - unbounded integer (i.e., `Num`)
  - fixed-width bitvector (i.e., `BV<N>`)

Reply:
> This code snippet makes me wonder why we are not fusing the `NumOperation` into `Type`.
> 
> Here is the reasoning:
> 
> * In Move Model, every expression has a `NodeId` which is associated with `Type`.
> * Now in this PR you are introducing a `NumOperation` and tries to associate some `NodeId` with it
> * A natural reaction will be, can we merge `NumOperation` with `Type`? Or, the other way round, whether a number is a bitvec or a math number is a _typing_ information, it makes perfect sense to me to merge them. So the question becomes: what is the concern that prevent you from merging them?
> 
> In fact, I am looking at the `Type` enum in Move Model and one way I see is the following:
> 
> ```diff
> pub enum PrimitiveType {
>     Bool,
>     U8,
>     U16,
>     U32,
>     U64,
>     U128,
>     U256,
>     Address,
>     Signer,
>     // Types only appearing in specifications
>     Num,
> +   BV8,
> +   BV16,
> +   BV32,
> +   BV64,
> +   BV128,
> +   BV256,
> +   NumOrBV // not sure whether we need this bottom case though
>     Range,
>     EventStore,
> }
> ```
> 
> This seems like a more elegant and compact representation of bitvec, plus you get to re-use most of the utilities we have around `Type`, e.g., type unification etc.

I agree that by design it is cleaner to merge `NumOperation` into `Type`. The current implementation touches the move-model as little as possible to speed up the implementation without bringing in more bugs. We can discuss more about refactor.